src/shogun/modelselection/parameter_tree.cpp

```diff
@@ -0,0 +1,473 @@
+{
+	auto param_iter = param.find(delimiter);
+
+	if (param_iter != std::string::npos)
+	{
+		return set_param_helper(
+		    param.substr(param_iter + delimiter.size()), node);
+	}
+	else
+	{
+		for (auto& node_i : m_nodes)
+			if (node_i.first == param)
+			{
+				node_i.second = node;
+				return true;
+			}
+	}
+
+	return false;
+}
+
+std::string ParameterNode::to_string() const
+{
+	std::stringstream ss;
+	std::unique_ptr<AnyVisitor> visitor(new ToStringVisitor(&ss));
+	ss << get_name() << "(" << get_parent_name() << "(";
+
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+
+	if (!m_nodes.empty())
+	{
+		ss << ", ";
+
+		for (auto node = m_nodes.begin(); node != m_nodes.end(); ++node)
+		{
+			ss << (*node).first << "=";
+			(*node).second->to_string_helper(ss, visitor);
+			if (std::next(node) != m_nodes.end())
+				ss << ", ";
+		}
+	}
+	ss << "))";
+	return ss.str();
+}
+
+void ParameterNode::to_string_helper(
+    std::stringstream& ss, std::unique_ptr<AnyVisitor>& visitor) const noexcept
+{
+	ss << get_parent_name() << "(";
+	for (auto param_pair = m_param_mapping.begin();
+	     param_pair != m_param_mapping.end(); ++param_pair)
+	{
+		ss << param_pair->first.c_str() << "=";
+		param_pair->second.visit(visitor.get());
+		if (std::next(param_pair) != m_param_mapping.end())
+			ss << ", ";
+	}
+	ss << ")";
+}
+
+bool ParameterNode::set_param_helper(const std::string& param, const Any& value)
+{
+	auto param_iter = param.find(delimiter);
+
+	if (param_iter != std::string::npos)
+	{
+		for (auto& node : m_nodes)
+		{
+			return node.second->set_param_helper(
+			    param.substr(param_iter + delimiter.size()), value);
+		}
+	}
+	else
+	{
+		if (m_parent->has(param))
+		{
+			m_param_mapping.insert(std::make_pair(param, value));
+			return true;
+		}
+	}
+
+	return false;
+}
+
+ParameterNode *ParameterNode::get_current() {
+	auto tree = new ParameterNode(*m_parent);
+
+	std::string param;
+
+	auto get_lambda = [&tree, &param](auto val) {
+		tree->attach(param, val);
+	};
+
+	for (auto& param_pair : m_param_mapping)
+	{
+		param = param_pair.first;
+		sg_any_dispatch(param_pair.second, sg_vector_typemap, shogun::None{}, get_lambda);
+	}
+	return tree;
+}
+
+CSGObject *ParameterNode::to_object(const std::shared_ptr<ParameterNode>& tree) {
+}
+
+ModelSelectionTree::ModelSelectionTree(CSGObject& model)
+    : first(true), m_counter(0), m_node_complete(false)
+{
+	m_parent = std::shared_ptr<CSGObject>(model.clone());
+	for (auto const& param : model.get_params())
+	{
+		if (auto* value = m_parent->get(param.first, std::nothrow))
+		{
+			create_node(param.first, value);
+			SG_UNREF(value);
+		}
+	}
+	m_current_node = m_nodes.begin();
+}
+
+void ModelSelectionTree::reset()
```

Comment:
what is that needed for?

Reply:
resets the internal state of a node, so that it goes back from the beginning when called from the parent node.